### ISSUE TITLE
EnsureIndex on a field in child documents

### ISSUE THREAD

**Issue by sanjeevbaral** [litedb-org/LiteDB#1553]
How do I create an index on a field that exists in the child document. For example, I have Patients collection and each patient has an array of Visit documents. I want to be able to search on a field that belongs to Visit using index. How do I do that? Not sure if I should use EnsureIndex("Visits[*].dateOfVisit")

**Comment by lbnascimento** [comment]
@sanjeevbaral Depends on what information you usually filter for in queries. For example, if you frequently search for patients with at least one visit since a specific date, you could run `EnsureIndex("MIN($.Visits[*].dateofVisit)")`. If you do so, a query like `patients.Find("MIN($.Visits[*].dateofVisit) > '2020-01-01'")` will use the index.

**Comment by sanjeevbaral** [comment]
Thanks and if I just wanted to find all visits on day or later than a day I would just use EnsureIndex("Visits[*].dateOfVisit")? The following would use the index? 

collection.EnsureIndex("Visits[*].dateOfVisit");
var patients = collection.Find("Visits[*].dateOfVisit ALL = '22020-01-01'");
int count = patients.Count();

**Comment by JKamsker** [comment]
Yes, `Visits[*].DateOfVisit` is the right index. It's a multikey index with one entry per visit (stored as `MAP($.Visits[*]=>@.DateOfVisit)`). The query needs two changes, though:

- use `ANY` ("at least one visit matches"), not `ALL` ("every visit matches"), and
- compare against a date value, not a string. `'2020-01-01'` is a string, and comparing a `DateTime` with a string doesn't compare dates. Pass the date as a parameter.

```csharp
col.EnsureIndex("visitDate", "$.Visits[*].DateOfVisit");

var since = new DateTime(2020, 1, 1);
var patients = col.Query()
    .Where("$.Visits[*].DateOfVisit ANY >= @0", since)
    .ToList();
```

The strongly typed version:

```csharp
col.EnsureIndex(p => p.Visits.Select(v => v.DateOfVisit));

var patients = col.Query()
    .Where(p => p.Visits.Select(v => v.DateOfVisit).Any(d => d >= since))
    .ToList();
```

`GetPlan()` shows an `INDEX SCAN(... >= ...)` for both. The query expression must match the index expression exactly, including field-name casing: an index on `Visits[*].dateOfVisit` is not used by a query on `DateOfVisit`, and the query falls back to a full scan. I checked this on 5.0.21 and the 6.x prerelease.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
